### PR TITLE
[log_monitor] Seek when reopening a file due to inode change

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

When reopening a file due to an inode change, we weren't seeking back to the right location. Now we are (with a unit test).

## Related issue number

Closes (but not really until it's cherry-picked) https://github.com/ray-project/ray/issues/27507

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Will do a bit more manual testing after wheels build too.

**Comment by wuisawesome** [comment]
Ok I ran some silly but complicated workload and it seems to work properly.

image

Notice there is no repeating of log messages (there are messages minutes apart from each other with nothing in between.

The details of the workload are unimportant, but i just wanted something that has some rapid messages, takes a while, etc, but for posterity, the workload was 

```
import ray

ray.init()
@ray.remote(num_cpus=16)
class Foo:
    pass

# Fill up the head node
actor = Foo.remote()
@ray.remote(num_cpus=16)
def foo():
    pass

@ray.remote(num_cpus=16)
class Actor:
    def run(self):
        ref = foo.remote()
        from time import sleep
        while True:
            finished, remaining = ray.wait([ref], timeout=0)
            if len(finished) > 0:
                print("done!")
                break
            print(f"remaining: {len(remaining)}, sleeping")
            sleep(5)

handle = Actor.remote()
ray.get(handle.run.remote())
```

File: python/ray/_private/log_monitor.py [hunk 1]
```diff
@@ -73,10 +73,13 @@ def reopen_if_necessary(self):
         would have different inodes, such as log rotation or file syncing
         semantics.
         """
-        open_inode = os.fstat(self.file_handle.fileno()).st_ino
+        open_inode = None
+        if self.file_handle is not None and not self.file_handle.closed:
```
Comment (ericl) [inline python/ray/_private/log_monitor.py:77]:
```suggestion
        if self.file_handle and not self.file_handle.closed:
             open_inode = os.fstat(self.file_handle.fileno()).st_ino
        else:
             open_inode = None
```

**Review by ericl — APPROVED** [review]

**Comment by wuisawesome** [comment]
test_error_handling looks flakey on master, everything else is looking good, merging

**Pull request merged**